### PR TITLE
MHK Tidal Plant

### PR DESCRIPTION
**Pull request by kbrunik**
# Add Marine Hydrokinetic Tidal Plant

Added MHKTidalPlant model to HOPP, modeling is based on MHKWavePlant model already integrated HOPP. This is the first of a series of PR's related to the tidal integration. The required first step is to add a resource module for tidal resource data and a MHKTidalPlant, which can be simulated as a standalone plant.

The MHKTidalPlant leverages [PySAM MhkTidal](https://nrel-pysam.readthedocs.io/en/main/modules/MhkTidal.html) model to simulate the MHK Tidal Plant performance. It leverages [PySAM MhkCosts](https://nrel-pysam.readthedocs.io/en/main/modules/MhkCosts.html) to calculate costs for the Reference Model 1 (tidal current turbine) from the [Reference Model Project](https://energy.sandia.gov/programs/renewable-energy/water-power/projects/reference-model-project-rmp/).

## PR Checklist

- [x] `RELEASE.md` has been updated to describe the changes made in this PR
- [x] Documentation
  - [x] Docstrings are up-to-date
  - [x] Related `docs/` files are up-to-date, or added when necessary
  - [x] Documentation has been rebuilt successfully
  - [ ] Examples have been updated
- [x] Tests pass (If not, and this is expected, please elaborate in the tests section)
- [x] PR description thoroughly describes the new feature, bug fix, etc.

## Impacted areas of the software

- `hopp/simulation/resource_files/tidal/Tidal_resource_timeseries.csv`
  - Tidal resource data that is used in the tests and for examples.
- `hopp/simulation/technologies/financial/mhk_cost_model.py`
  - Raise a `ValueError` if the correct reference model is not selected
  - Assign `marine_energy_tech` appropriately if it's a tidal device or WEC.
- `hopp/simulation/technologies/resource/tidal_resource.py`
  - Load in Tidal Resource data and format it for MHKTidalPlant performance model input
- `hopp/simulation/technologies/sites/site_info.py`
  - Add `TidalResource` to `SiteInfo`
- `hopp/simulation/technologies/tidal/mhk_tidal_plant.py`
  - Add class `MHKTidalPlant` which inherits from `PowerSource`
  - Set necessary attributes to run PySAM MhkTidal model
  - `create_mhk_cost_calculator()` instantiates the cost model for the tidal plant
  - `calculate_total_installed_cost()` calculates the total installed cost
  - `system_capacity_by_num_devices()` adjusts system capacity based on number of devices
  - `simulate()` calculates the total installed cost and runs the performance simulation
- `tests/hopp/inputs/tidal/tidal_device.yaml`
  - An input YAML config based on the reference model 1 tidal current turbine
- `tests/hopp/test_tidal.py`
  - Test suite for `MHKTidalPlant`, `TidalResource`, `MHKCosts`

### CODE REVIEW

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+@define
+class MHKTidalConfig(BaseClass):
+    """
+    Configuration class for MHKTidalPlant.
+
+    Args:
+        device_rating_kw: Rated power of the MHK device [kW]
+        num_devices: Number of MHK tidal devices in the system
+        tidal_power_curve: Power curve of tidal energy device as function of stream speeds [kW]
+        tidal_resource: DUMMY VARIABLE. This is required by PySAM MhkTidal module even during timeseries simulations.
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:22]: How should I interpret this "DUMMY VARIABLE"? Do I always need to set it or is it sometimes not used? Slightly more context here would be great

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        fin_model: Optional financial model. Can be any of the following:
+            - a dict representing a `CustomFinancialModel`
+            - an object representing a `CustomFinancialModel` instance
+        loss_array_spacing: Array spacing loss in % (default: 0)
+        loss_resource_overprediction: Resource overprediction loss
+            in % (default: 0)
+        loss_transmission: Transmission loss in % (default: 0)
+        loss_downtime: Array/WEC downtime loss in % (default: 0)
+        loss_additional: Additional losses in % (default: 0)
+        layout_mode: TODO
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:33]: I assume this is something in the next PR or two? Or is it meant to be part of this one?

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    """
+    Marine Hydrokinetic (MHK) Tidal Plant.
+
+    Args:
+        site: Site information
+        config: MHK system configuration parameters
+        cost_model_inputs: An optional dictionary containing input parameters for
+            cost modeling.
+
+        """
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:59]: dedent

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        
+        if self.cost_model_inputs is not None:
+            self.mhk_costs = MHKCosts(self.config, self.cost_model_inputs)
+        else:
+            self.mhk_costs = None
+            
+        super().__init__("MHKTidalPlant", self.site, system_model, financial_model)
+
+        # Set tidal resource model choice
+        system_model.MHKTidal.tidal_resource_model_choice = 1 # Time-series data=1 JPD=0
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:85]: do I need to know what JPD=0 means here?

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            'loss_transmission', 
+            'loss_additional'
+            ]
+
+        for attribute in loss_attributes:
+            if attribute in self.config.as_dict().keys():
+                attr = getattr(self.config, attribute)
+                setattr(self._system_model.MHKTidal, attribute, attr)
+            else:
+                setattr(self._system_model.MHKTidal, attribute, 0)
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:112]:
```suggestion
        for attribute in loss_attributes:
            attr = getattr(self.config, attribute, 0)
            setattr(self._system_model.MHKTidal, attribute, attr)
```

Comment (johnjasa) [reply hopp/simulation/technologies/tidal/mhk_tidal_plant.py:112]: You can provide a default for `getattr` to slightly simplify this: https://docs.python.org/3/library/functions.html#getattr

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        
+        return self._financial_model.value("total_installed_cost", total_installed_cost)
+
+    def system_capacity_by_num_devices(self, tidal_size_kw: float):
+        """
+        Sets the system capacity by adjusting the number of devices
+        """
+        new_num_devices = round(tidal_size_kw / self.device_rated_power)
+        if self.number_devices != new_num_devices:
+            self.number_devices = new_num_devices
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:163]:
```suggestion
        self.number_devices = new_num_devices
```

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            self.mhk_costs.device_rated_power = device_rate_power
+
+    @property
+    def number_devices(self) -> int:
+        return self._system_model.MHKTidal.number_devices
+
+    @number_devices.setter
+    def number_devices(self, number_devices: int):
+        self._system_model.MHKTidal.number_devices = number_devices
+        if self.mhk_costs != None:
```
Comment (johnjasa) [inline hopp/simulation/technologies/tidal/mhk_tidal_plant.py:199]: Non-blocking nitpick that requires no change here: I'd suggest using `is not None` instead of `!= None` as None is a singleton. Not that important and using `!=` doesn't usually have negative impacts, but I was caught by an edge case once so I like to suggest it. https://stackoverflow.com/a/2209781/5050416 for more info 
(I know this occurs throughout HOPP but seeing it here made me want to mention it)

File: tests/hopp/inputs/tidal/tidal_device.yaml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+- [3.000000, 1115.000000]
+- [3.100000, 1115.000000]
+- [3.200000, 1085.370000]
+- [3.300000, 1055.730000]
+
+num_devices: 20
+# Tidal resource is required in PySAM prechecks
+# this is a dummy resource profile and does not 
+# impact simulation when using timeseries data
+# TODO: Remove once PySAM Pypi updates
```
Comment (johnjasa) [inline tests/hopp/inputs/tidal/tidal_device.yaml:52]: Is this update to PySAM Pypi coming up imminently, was it already fixed, or something else?

File: tests/hopp/test_tidal.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+@fixture
+def site():
+	data = {
+		"lat": 44.6899,
+		"lon": 124.1346,
+		"year": 2010,
+		"tz": -7,
+	}
+	tidal_resource_file = ROOT_DIR / "simulation" / "resource_files" / "tidal" / "Tidal_resource_timeseries.csv"
```
Comment (johnjasa) [inline tests/hopp/test_tidal.py]: Non-blocking suggestion that shouldn't be changed here (because this is used throughout HOPP) but for future reference: I'd suggest using the Path package and its `.join()` method in the future; it's a bit more portable for other systems

**Review by johnjasa — CHANGES_REQUESTED** [review]
Awesome progress here! I really like how you're doing this in staged PRs with a clear demarcation between the capabilities added.

I made some small comments and suggestions. In places where no change is required I tried to make that clear, but also give feedback that might be useful when developing code together in the future. I also directly pushed a repo-wide conversion for tabs to spaces (only occurred in five files) that I noticed as part of this PR.

I'm "requesting changes" but really nothing major, I just want to make sure I come back to see your responses to some of these small questions. Looking forward to the next PRs!

File: hopp/simulation/technologies/financial/mhk_cost_model.py [hunk 10]
```diff
@@ -75,7 +75,11 @@ def __attrs_post_init__(self):
         self._device_spacing = self.cost_model_inputs.device_spacing
         self._cable_sys_overbuild = self.cost_model_inputs.cable_system_overbuild
         
-        self._ref_model_num = "RM"+str(self.cost_model_inputs.reference_model_num)
+        ref_model_numbers = {1,3,5,6}
+        if self.cost_model_inputs.reference_model_num in ref_model_numbers:
+            self._ref_model_num = f"RM{self.cost_model_inputs.reference_model_num}"
+        else:
+            raise ValueError("reference_model_num can be 1, 3, 5 or 6")
```
[the remark this answers is not in the document]
Comment (kbrunik) [reply hopp/simulation/technologies/financial/mhk_cost_model.py:82]: I can certainly try to add a bit more info about the reference models but they're THE models for marine energy.

File: tests/hopp/inputs/tidal/tidal_device.yaml (hunk 8 above)
Comment (kbrunik) [reply tests/hopp/inputs/tidal/tidal_device.yaml:52]: PySAM has this captured in an [issue](https://github.com/NREL/ssc/issues/1279), based on talking with the PySAM developer the fix should be soon, but I should be notified when the issue is resolved on GitHub. Is there a different way you'd like to hand this?

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py (hunk 2 above)
Comment (kbrunik) [reply hopp/simulation/technologies/tidal/mhk_tidal_plant.py:33]: Removed - carry over from how I implanted in wave modeling.

File: hopp/simulation/technologies/tidal/mhk_tidal_plant.py (hunk 1 above)
Comment (kbrunik) [reply hopp/simulation/technologies/tidal/mhk_tidal_plant.py]: Check out the new Docstring description and let me know if that's clearer

File: tests/hopp/test_tidal.py (hunk 9 above)
Comment (kbrunik) [reply tests/hopp/test_tidal.py]: I used `.joinpath()` from Path, let me know if the change is what you were intending.

File: tests/hopp/inputs/tidal/tidal_device.yaml (hunk 8 above)
Comment (johnjasa) [reply tests/hopp/inputs/tidal/tidal_device.yaml:52]: Nope, that's great! I wasn't sure about the timeline etc, this is perfect.

File: hopp/simulation/technologies/financial/mhk_cost_model.py [hunk 11]
```diff
@@ -16,15 +16,29 @@ class MHKCostModelInputs(BaseClass):
...
+    Note:
+        More information about the reference models and their 
+        associated costs can be found in the 
+        [Reference Model Project](https://energy.sandia.gov/programs/renewable-energy/water-power/projects/reference-model-project-rmp/)
+        
+        The supported reference models in this cost model are:
+            - Reference Model 1: Tidal Current Turbine
+            - Reference Model 3: Wave Point Absorber
+            - Reference Model 5: Oscillating Surge Flap
+            - Reference Model 6: Oscillating Water Column
```
Comment (johnjasa) [inline hopp/simulation/technologies/financial/mhk_cost_model.py:38]: Very cool, love this detail in the docstring!

**Review by johnjasa — APPROVED** [review]
